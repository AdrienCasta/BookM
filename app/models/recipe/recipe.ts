import { Instance, SnapshotOut, types } from "mobx-state-tree"
import * as yup from "yup"
import type { Asserts } from "yup"

export interface IRecipeFieldValues extends Asserts<typeof RecipeSchema> {}

export const RecipeSchema = yup.object({
  title: yup.string().required(),
  image: yup.object({ uri: yup.string().required() }).required(),
  description: yup.string().required(),
  numberOfPersons: yup.number().positive().integer().required(),
  time: yup.number().positive().integer().required(),
  steps: yup
    .array(yup.object({ description: yup.string().required(), trick: yup.string() }))
    .min(2),
  ingredients: yup
    .array(
      yup.object({
        image: yup.object({ uri: yup.string().required() }).required(),
        label: yup.string().required(),
      }),
    )
    .transform((value) => value.filter(({ label, image }) => label && image))
    .min(2)
    .required(),
  cookingTime: yup
    .number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .positive()
    .integer()
    .notRequired(),
  numberOfCalories: yup
    .number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .positive()
    .integer()
    .notRequired(),
})

/**
 * Model description here for TypeScript hints.
 */
export const RecipeModel = types.model({
  title: types.string,
  description: types.string,
  image: types.model({ uri: types.string }),
  numberOfPersons: types.number,
  time: types.number,
  ingredients: types.array(
    types.model({ label: types.string, image: types.model({ uri: types.string }) }),
  ),
  cookingTime: types.union(types.number, types.number),
  numberOfCalories: types.union(types.number, types.number),
  steps: types.array(
    types.model({ description: types.string, trick: types.union(types.string, types.null) }),
  ),
})

/**
  * Un-comment the following to omit model attributes from your snapshots (and from async storage).
  * Useful for sensitive data like passwords, or transitive state like whether a modal is open.

  * Note that you'll need to import `omit` from ramda, which is already included in the project!
  *  .postProcessSnapshot(omit(["password", "socialSecurityNumber", "creditCardNumber"]))
  */

type RecipeType = Instance<typeof RecipeModel>
export interface Recipe extends RecipeType {}
type RecipeSnapshotType = SnapshotOut<typeof RecipeModel>
export interface RecipeSnapshot extends RecipeSnapshotType {}
