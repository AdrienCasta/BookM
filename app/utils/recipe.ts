import UserIcon from "../screens/recipe-creation/assets/user.svg"
import PanIcon from "../screens/recipe-creation/assets/pan.svg"
import StopWatchIcon from "../screens/recipe-creation/assets/stopwatch.svg"
import FireIcon from "../screens/recipe-creation/assets/fire.svg"
import { IRecipeFieldValues } from "../models/recipe/recipe"
import { FC } from "react"
import { SvgProps } from "react-native-svg"

export const recipeQuantifiablelIcons = new Map<
  keyof Pick<IRecipeFieldValues, "numberOfPersons" | "time" | "numberOfCalories" | "cookingTime">,
  FC<SvgProps>
>([
  ["numberOfPersons", UserIcon],
  ["time", StopWatchIcon],
  ["cookingTime", PanIcon],
  ["numberOfCalories", FireIcon],
])
