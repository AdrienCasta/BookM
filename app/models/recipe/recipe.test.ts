import { getSnapshot } from "mobx-state-tree"
import { RecipeStore } from "./recipe"

const mockRecipe = {
  numberOfCalories: null,
  cookingTime: null,
  ingredients: [
    {
      label: "",
      image: {
        uri: "test",
      },
    },
  ],
  steps: [
    {
      description: "",
      trick: "",
    },
    {
      description: "",
      trick: "",
    },
  ],
  time: new Date(0),
  numberOfPersons: 0,
  description: "",
  image: {
    uri: "",
  },
  title: "",
}

jest.mock("aws-amplify", () => {
  return {
    API: {
      graphql: jest.fn().mockImplementation(() => {
        return Promise.resolve({
          data: {
            listRecipes: {
              items: [
                { ...mockRecipe, id: 1 },
                { ...mockRecipe, id: 2 },
              ],
            },
          },
        })
      }),
    },
  }
})

test("can be created", () => {
  const instance = RecipeStore.create({})

  expect(instance).toBeTruthy()
})

it("recipe store can add new recipe", () => {
  const recipe = RecipeStore.create({})
  recipe.addRecipe(mockRecipe)
  expect(getSnapshot(recipe)).toMatchSnapshot()
})
it("recipe store can clear recipe", () => {
  const recipe = RecipeStore.create(mockRecipe)
  recipe.createRecipe()
  expect(getSnapshot(recipe)).toMatchSnapshot()
})
it("recipe store can list recipes", () => {
  const recipe = RecipeStore.create(mockRecipe)
  recipe.listRecipes()
  expect(getSnapshot(recipe)).toMatchSnapshot()
})
