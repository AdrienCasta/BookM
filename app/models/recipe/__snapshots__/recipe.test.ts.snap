// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipe store can add new recipe 1`] = `
Object {
  "recipe": Object {
    "cookingTime": null,
    "description": "",
    "image": Object {
      "uri": "",
    },
    "ingredients": Array [
      Object {
        "image": Object {
          "uri": "test",
        },
        "label": "",
      },
    ],
    "numberOfCalories": null,
    "numberOfPersons": 0,
    "steps": Array [
      Object {
        "description": "",
        "trick": "",
      },
      Object {
        "description": "",
        "trick": "",
      },
    ],
    "time": 0,
    "title": "",
  },
  "recipes": Array [],
}
`;

exports[`recipe store can clear recipe 1`] = `
Object {
  "recipe": undefined,
  "recipes": Array [],
}
`;

exports[`recipe store can list recipes 1`] = `
Object {
  "recipe": undefined,
  "recipes": Array [],
}
`;
